File: sandbox/screen-reading/shared.lua
-- Shared command area capture logic for screen-reading experiments.
--
-- The command area begins at comStart. It holds 4 text rows.
-- Each text row is 32 characters wide and 8 scanlines tall.
-- A scanline is 32 bytes wide — one byte per character position.
-- Successive scanlines are 32 bytes apart.
--
-- Which scanlines belong to which text row:
--   Text row 1: first 8 scanlines
--   Text row 2: next 8 scanlines
--   Text row 3: next 8 scanlines
--   Text row 4: last 8 scanlines
--
-- Example: the 4th character in the 2nd text row.
--   The 2nd row begins at the 9th scanline, 256 bytes from comStart
--   (8 preceding scanlines × 32 bytes each).
--   The 4th character starts 3 bytes into that scanline.
--   Top pixel row of the character:           comStart + 256 + 3
--   Next pixel row (one scanline down):        comStart + 288 + 3
--   This repeats for all 8 scanlines of the text row.
--
-- For the code, scanlines and character positions are zero-based.
-- Address for scanline s, character position c:  comStart + s × 32 + c
--
--
-- The capture function logs the command area pixels to a file.
-- It only writes a frame when the pixels differ from the previous
-- logged frame. Identical frames are tracked and summarized as a
-- single UNCHANGED line.
--
-- Output format:
--   FRAME,<frame>,<areaStart>,<areaSize>,<cursor>,<color>,<displayFn>,<match>
--   DATA,<scanline>,<32 hex bytes>
--   ...
--   UNCHANGED,<firstFrame>,<lastFrame>
--
-- The entire frame is built in memory (table.concat) and written in
-- a single call to avoid hitting MAME's io.write buffer limit.

local TEXT_ROWS = 4
local CHARS_PER_ROW = 32
local SCANLINES_PER_ROW = 8
local TOTAL_SCANLINES = TEXT_ROWS * SCANLINES_PER_ROW

local commandAreaSnapshot = {
    pixels = nil,   -- 1024 bytes from the most recent logged frame
    frame = nil,    -- the frame number when those pixels appeared
}

local function _readPixels(memory)
    local areaStart = memory:read_u8(0x0390) * 256 + memory:read_u8(0x0391)
    local pixels = {}
    for scanlineIndex = 0, TOTAL_SCANLINES - 1 do
        local scanlineStart = areaStart + scanlineIndex * CHARS_PER_ROW
        for characterColumn = 0, CHARS_PER_ROW - 1 do
            pixels[#pixels + 1] = memory:read_u8(scanlineStart + characterColumn)
        end
    end
    return pixels
end

local function _pixelsMatch(current, previous)
    if #current ~= #previous then
        return false
    end
    for i = 1, #current do
        if current[i] ~= previous[i] then
            return false
        end
    end
    return true
end

local function _writeFrame(logFile, frame, memory)
    local areaStart = memory:read_u8(0x0390) * 256 + memory:read_u8(0x0391)
    local areaSize   = memory:read_u8(0x0392) * 256 + memory:read_u8(0x0393)
    local cursor     = memory:read_u8(0x0394) * 256 + memory:read_u8(0x0395)
    local color      = memory:read_u8(0x0396)
    local displayFn  = memory:read_u8(0x02B2) * 256 + memory:read_u8(0x02B3)
    local match      = memory:read_u8(0x027B)

    local logPieces = {string.format("FRAME,%d,%d,%d,%d,%d,%d,%d\n",
        frame, areaStart, areaSize, cursor, color, displayFn, match)}

    for scanlineIndex = 0, TOTAL_SCANLINES - 1 do
        local scanlineStart = areaStart + scanlineIndex * CHARS_PER_ROW
        logPieces[#logPieces + 1] = string.format("DATA,%d,", scanlineIndex)

        for characterColumn = 0, CHARS_PER_ROW - 1 do
            local byte = memory:read_u8(scanlineStart + characterColumn)
            logPieces[#logPieces + 1] = string.format("%02X", byte)
        end

        logPieces[#logPieces + 1] = "\n"
    end

    logFile:write(table.concat(logPieces))
    logFile:flush()
end

local function captureCommandAreaPixels(logFile, frame, memory)
    local currentPixels = _readPixels(memory)

    if commandAreaSnapshot.pixels and _pixelsMatch(currentPixels, commandAreaSnapshot.pixels) then
        -- Screen has not changed. Nothing to write.
        return
    end

    -- Screen changed. Summarize any identical frames we skipped,
    -- then write the new frame and update the snapshot.
    if commandAreaSnapshot.frame then
        local skippedFrames = frame - commandAreaSnapshot.frame - 1
        if skippedFrames > 0 then
            logFile:write(string.format("UNCHANGED,%d,%d\n",
                commandAreaSnapshot.frame + 1, frame - 1))
            logFile:flush()
        end
    end

    _writeFrame(logFile, frame, memory)
    commandAreaSnapshot.pixels = currentPixels
    commandAreaSnapshot.frame = frame
end

return {
    captureCommandAreaPixels = captureCommandAreaPixels,
}

File: sandbox/screen-reading/shared.py
"""Shared utilities for screen-reading experiments.

Font patterns extracted from docs/references/game/code.md, TextCharacters section.
Each character has 7 rows of 5 bits. The leftmost pixel is the most significant bit.
"""

# Reference patterns extracted from code.md (32 characters: space, A-Z, !, _, ?, .)
FONT_PATTERNS = [
    (" ", [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]),
    ("A", [0x04, 0x0A, 0x11, 0x11, 0x1F, 0x11, 0x11]),
    ("B", [0x1E, 0x11, 0x11, 0x1E, 0x11, 0x11, 0x1E]),
    ("C", [0x0E, 0x11, 0x10, 0x10, 0x10, 0x11, 0x0E]),
    ("D", [0x1E, 0x11, 0x11, 0x11, 0x11, 0x11, 0x1E]),
    ("E", [0x1F, 0x10, 0x10, 0x1E, 0x10, 0x10, 0x1F]),
    ("F", [0x1F, 0x10, 0x10, 0x1E, 0x10, 0x10, 0x10]),
    ("G", [0x0F, 0x11, 0x10, 0x10, 0x13, 0x11, 0x0F]),
    ("H", [0x11, 0x11, 0x11, 0x1F, 0x11, 0x11, 0x11]),
    ("I", [0x0E, 0x04, 0x04, 0x04, 0x04, 0x04, 0x0E]),
    ("J", [0x01, 0x01, 0x01, 0x01, 0x01, 0x11, 0x0E]),
    ("K", [0x11, 0x12, 0x14, 0x18, 0x14, 0x12, 0x11]),
    ("L", [0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x1F]),
    ("M", [0x11, 0x1B, 0x15, 0x15, 0x15, 0x11, 0x11]),
    ("N", [0x11, 0x11, 0x19, 0x15, 0x13, 0x11, 0x11]),
    ("O", [0x0E, 0x11, 0x11, 0x11, 0x11, 0x11, 0x0E]),
    ("P", [0x1E, 0x11, 0x11, 0x1E, 0x10, 0x10, 0x10]),
    ("Q", [0x0E, 0x11, 0x11, 0x11, 0x15, 0x12, 0x0D]),
    ("R", [0x1E, 0x11, 0x11, 0x1E, 0x14, 0x12, 0x11]),
    ("S", [0x0E, 0x11, 0x10, 0x0E, 0x01, 0x11, 0x0E]),
    ("T", [0x1F, 0x15, 0x04, 0x04, 0x04, 0x04, 0x04]),
    ("U", [0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x0E]),
    ("V", [0x11, 0x11, 0x11, 0x0A, 0x0A, 0x04, 0x04]),
    ("W", [0x11, 0x11, 0x11, 0x15, 0x15, 0x1B, 0x11]),
    ("X", [0x11, 0x11, 0x0A, 0x04, 0x0A, 0x11, 0x11]),
    ("Y", [0x11, 0x11, 0x0A, 0x04, 0x04, 0x04, 0x04]),
    ("Z", [0x1F, 0x01, 0x02, 0x04, 0x08, 0x10, 0x1F]),
    ("!", [0x04, 0x04, 0x04, 0x04, 0x04, 0x00, 0x04]),
    ("_", [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x1F]),
    ("?", [0x0E, 0x11, 0x01, 0x06, 0x04, 0x00, 0x04]),
    (".", [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04]),
]


def load_frames(path):
    """Parse log.txt into list of (metadata, scanlines) tuples."""
    with open(path) as f:
        lines = [l.strip() for l in f]

    frames = []
    scanlines = []
    meta = None
    for line in lines:
        if line.startswith("FRAME,"):
            if meta and len(scanlines) == 32:
                frames.append((meta, list(scanlines)))
            parts = line.split(",")
            meta = {
                "frame": int(parts[1]),
                "area_start": int(parts[2]),
                "area_size": int(parts[3]),
                "cursor": int(parts[4]),
                "color": int(parts[5]),
                "display_fn": int(parts[6]),
                "pm": int(parts[7]),
            }
            scanlines = []
        elif line.startswith("DATA,"):
            parts = line.split(",", 2)
            if len(parts) >= 3:
                scanlines.append(parts[2])
    if meta and len(scanlines) == 32:
        frames.append((meta, scanlines))
    return frames


def decode_character(screen_bytes, char_offset, com_color):
    """Read 7 bytes at 32-byte intervals (one per scanline) and match against font."""
    candidate = []
    for row in range(7):
        addr = char_offset + row * 32
        if addr >= len(screen_bytes):
            return "?"
        byte = screen_bytes[addr]
        row_val = (byte ^ com_color) >> 2 & 0x1F
        candidate.append(row_val)

    best_char = "?"
    best_dist = 999

    for char, pattern in FONT_PATTERNS:
        dist = 0
        for r in range(7):
            diff = candidate[r] ^ pattern[r]
            dist += diff.bit_count()
        if dist < best_dist:
            best_dist = dist
            best_char = char

    if best_dist > 8:
        return "?"
    return best_char


def decode_text_row(screen_bytes, text_row, com_color, characters_per_row=32):
    """Decode one text row. text_row 0 = first row in the command area."""
    start_byte = text_row * 8 * characters_per_row
    result = []
    for col in range(characters_per_row):
        ch = decode_character(screen_bytes, start_byte + col, com_color)
        result.append(ch)
    return "".join(result)


def decode_frame(screen_bytes, com_size, com_color):
    """Decode all 32 character positions in the first text row. Legacy wrapper."""
    return decode_text_row(screen_bytes, 0, com_color, min(com_size, 32))
